docs/tutorial/native-code-and-electron.md

```diff
@@ -0,0 +1,375 @@
+
+# Tutorial: Creating a Native Node.js Addon for Electron
+
+This tutorial will walk you through building a basic Node.js native addon that can be used in Electron applications. We'll focus on concepts common to all platforms, using C++ as the implementation language. Once you complete this tutorial common to all native Node.js addons, you can move on to one of our platform-specific tutorials.
+
+This tutorial assumes you have Node.js and npm installed, as well as the basic tools necessary for compiling code on your platform (like Visual Studio on Windows, Xcode on macOS, or GCC/Clang on Linux). You can find detailed instructions in the [`node-gyp` readme](https://github.com/nodejs/node-gyp?tab=readme-ov-file).
+
+### Requirements: macOS
+
+To build native Node.js addons on macOS, you'll need the Xcode Command Line Tools. These provide the necessary compilers and build tools (namely, `clang`, `clang++`, and `make`). The following command will prompt you to install the Command Line Tools if they aren't already installed.
+
+```sh
+xcode-select --install
+```
+
+### Requirements: Windows
+
+The official Node.js installer offers the optional installation of "Tools for Native Modules", which installs everything required for the basic compilation of C++ modules - specifically, Python 3 and the "Visual Studio Desktop development with C++" workload. Alternatively, you can use `chocolatey`, `winget`, or the Windows Store.
+
+### Requirements: Linux
+
+* [A supported version of Python](https://devguide.python.org/versions/)
+* `make`
+* A proper C/C++ compiler toolchain, like [GCC](https://gcc.gnu.org)
+
+## 1) Creating a package
+
+First, create a new Node.js package that will contain your native addon:
+
+```sh
+mkdir my-native-addon
+cd my-native-addon
+npm init -y
+```
+
+This creates a basic `package.json` file. Next, we'll install the necessary dependencies:
+
+```sh
+npm install --save node-addon-api bindings
+```
+
+* `node-addon-api`: This is a C++ wrapper for the low-level Node.js API that makes it easier to build addons. It provides a C++ object-oriented API that's more convenient and safer to use than the raw C-style API.
+* `bindings`: A helper module that simplifies the process of loading your compiled native addon. It handles finding your compiled .node file automatically.
+
+Now, let's update our package.json to include the appropriate build scripts. We will explain what these specifically do further below.
+
+```json
+{
+  "name": "my-native-addon",
+  "version": "1.0.0",
+  "description": "A native addon for Electron",
+  "main": "js/index.js",
+  "scripts": {
+    "clean": "rm -rf build",
+    "build": "node-gyp configure && node-gyp build"
```

Comment:
if you use node-gyp here, do we need to add in devDependencies node-gyp too?

Reply:
`node-gyp` comes with Node.js

Reply:
I _thought_ it comes with Node.js but I guess I was wrong. I'll add it!